src/transformers/models/llama/modeling_llama.py

```diff
@@ -314,15 +360,34 @@ def forward(
             key_states = self.k_proj(hidden_states)
             value_states = self.v_proj(hidden_states)
 
+        # Apply 2D sharding:
+        # query_states (batch, length, hidden)
+        # key_states (batch, length, hidden / attention_heads * key_value_heads)
+        # value_states (batch, length, hidden / attention_heads * key_value_heads)
+        # mesh (data, None, model)
+        data_model_mesh = xs.HybridMesh(ici_mesh_shape=(self.spmd_data_axis, 1, self.spmd_model_axis))
+        if self.spmd_debug:
+            print('> Sharding query_states', query_states.shape)
+            print('> Sharding key_states', key_states.shape)
+            print('> Sharding value_states', value_states.shape)
+        xs.mark_sharding(query_states, data_model_mesh, range(len(query_states.shape)))
+        xs.mark_sharding(key_states, data_model_mesh, range(len(key_states.shape)))
+        xs.mark_sharding(value_states, data_model_mesh, range(len(value_states.shape)))
+        if self.spmd_debug:
+            print(torch_xla._XLAC._get_xla_sharding_spec(query_states))
+            print(torch_xla._XLAC._get_xla_sharding_spec(key_states))
+            print(torch_xla._XLAC._get_xla_sharding_spec(value_states))
+
         query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
         key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
         value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
 
         kv_seq_len = key_states.shape[-2]
         if past_key_value is not None:
             kv_seq_len += past_key_value[0].shape[-2]
-        cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
-        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
+        # TODO: Disable rotary embedding for now.
+        # cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
+        # query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
```

Comment:
Is this needed for 2d sharding?

Reply:
I don't know what's the best way to shard rotary embedding. Just disabled for now.